### PR TITLE
[WIP] stopBuild/buildrequests Poc

### PR DESCRIPTION
**Pull request by tardyp**
This PR is the results of a brainstorming with my team on how to do the stopBuild properly.

Something that would also stop triggered builds, as we are using a lot of them.

The goal was to prepare a story for the next sprint.
I ended up practically writting most of the code. They still have the action to finish it to ramp up on the nine core.

### CODE REVIEW

File: master/buildbot/data/buildrequests.py [hunk 1]
```diff
@@ -98,6 +98,24 @@ def startConsuming(self, callback, options, kwargs):
...
+        # cancel the build anyway
+        try:
+            yield self.master.data.updates.claimBuildRequests([self.id])
+        except buildrequests.AlreadyClaimedError:
+            builds = yield self.master.data.get(("buildrequests", brid, "builds"))
+            for b in builds:
+                self.master.mq.produce(("control", "builds", b['buildid'], "stop"))
+            return
+
+        # then complete it with 'FAILURE'; this is the closest we can get to
```
Comment (sa2ajj) [inline master/buildbot/data/buildrequests.py]: what about EXCEPTION?

File: master/buildbot/data/buildrequests.py [hunk 2]
```diff
@@ -98,6 +98,24 @@ def startConsuming(self, callback, options, kwargs):
...
+    @defer.inlineCallbacks
+    def control(self, method, kwargs):
+        # first, try to claim the request; if this fails, then it's too late to
+        # cancel the build anyway
+        try:
+            yield self.master.data.updates.claimBuildRequests([self.id])
+        except buildrequests.AlreadyClaimedError:
+            builds = yield self.master.data.get(("buildrequests", brid, "builds"))
+            for b in builds:
+                self.master.mq.produce(("control", "builds", b['buildid'], "stop"))
```
Comment (sa2ajj) [inline master/buildbot/data/buildrequests.py]: so regardless the `method` it's always `stop`?

File: master/buildbot/data/buildrequests.py (hunk 1 above)
Comment (tardyp) [inline master/buildbot/data/buildrequests.py]: Indeed that should be CANCELED

File: master/buildbot/data/buildrequests.py [hunk 3]
```diff
@@ -98,6 +98,24 @@ def startConsuming(self, callback, options, kwargs):
         return self.master.mq.startConsuming(callback,
                                              ('buildrequests', None, None, None, None))
 
+    @defer.inlineCallbacks
+    def control(self, method, kwargs):
+        # first, try to claim the request; if this fails, then it's too late to
+        # cancel the build anyway
```
Comment (djmitche) [inline master/buildbot/data/buildrequests.py]: This should verify that `method == 'cancel'`

File: master/buildbot/data/buildrequests.py (hunk 2 above)
Comment (djmitche) [inline master/buildbot/data/buildrequests.py]: (sorry for my duplicate comment saying what @sa2ajj said.. /me backs away slowly)

Comment (sa2ajj) [inline master/buildbot/data/buildrequests.py]: com' on :)

File: master/buildbot/data/buildrequests.py [hunk 4]
```diff
@@ -98,6 +98,24 @@ def startConsuming(self, callback, options, kwargs):
...
+        try:
+            yield self.master.data.updates.claimBuildRequests([self.id])
+        except buildrequests.AlreadyClaimedError:
+            builds = yield self.master.data.get(("buildrequests", brid, "builds"))
```
Comment (tardyp) [inline master/buildbot/data/buildrequests.py]: there is a race condition here between claim and creation of build.

Comment (tardyp) [inline master/buildbot/data/buildrequests.py]: need to add a comment about it, as I dont have any suggestion on how to manage this race condition.

**Comment by tardyp** [comment]
superseeded by https://github.com/buildbot/buildbot/pull/1656

**Pull request closed without merging**